kube/src/client/discovery.rs

```diff
@@ -0,0 +1,341 @@
+use crate::{api::GroupVersionKind, Client};
+use k8s_openapi::apimachinery::pkg::apis::meta::v1::{APIResource, APIResourceList};
+use std::{cmp::Reverse, collections::HashMap};
+
+struct GroupVersionData {
+    version: String,
+    list: APIResourceList,
+    resources: Vec<APIResource>,
+}
+
+impl GroupVersionData {
+    fn new(version: String, list: APIResourceList) -> Self {
+        GroupVersionData {
+            version,
+            list: list.clone(),
+            resources: filter_api_resource_list(list),
+        }
+    }
+}
+
+/// Describes one API group.
+pub struct Group {
+    name: String,
+    versions_and_resources: Vec<GroupVersionData>,
+    preferred_version: Option<String>,
+}
+
+/// High-level utility for runtime API discovery.
+///
+/// On creation `Discovery` queries Kubernetes API,
+/// making list of all API resources, and provides a simple
+/// interface on the top of that information.
+pub struct Discovery {
+    groups: HashMap<String, Group>,
+}
+
+fn filter_api_resource_list(resource_list: APIResourceList) -> Vec<APIResource> {
+    let mut resource_list = resource_list.resources;
+    // skip subresources
+    resource_list.retain(|ar| !ar.name.contains('/'));
```

Comment:
Oh, we are discovering subresources like they are actual resources?

Reply:
Yeah, subresources are returned too, e.g.

```
{
      "name": "statefulsets/scale",
      "singularName": "",
      "namespaced": true,
      "group": "autoscaling",
      "version": "v1",
      "kind": "Scale",
      "verbs": [
        "get",
        "patch",
        "update"
      ]
    },
}
```